File: restrobazaar/lib/src/core/shipping.dart
/// Default shipping tiers (aligned with web Store Settings).
class ShippingTier {
  const ShippingTier({
    required this.minAmount,
    this.maxAmount,
    required this.charge,
  });

  final double minAmount;
  final double? maxAmount; // null = no upper limit
  final double charge;

  factory ShippingTier.fromJson(Map<String, dynamic> json) {
    return ShippingTier(
      minAmount: _toDouble(json['minAmount'], 0),
      maxAmount: json['maxAmount'] == null
          ? null
          : _toDouble(json['maxAmount'], 0),
      charge: _toDouble(json['charge'], 0),
    );
  }

  static double _toDouble(dynamic value, double fallback) {
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? fallback;
    return fallback;
  }
}

class ShippingSettings {
  const ShippingSettings({
    this.enabled = true,
    this.tiers = const [],
  });

  final bool enabled;
  final List<ShippingTier> tiers;

  static final defaults = ShippingSettings(
    enabled: true,
    tiers: const [
      ShippingTier(minAmount: 0, maxAmount: 2999, charge: 250),
      ShippingTier(minAmount: 3000, maxAmount: 5999, charge: 150),
      ShippingTier(minAmount: 6000, maxAmount: null, charge: 0),
    ],
  );

  factory ShippingSettings.fromJson(Map<String, dynamic>? json) {
    if (json == null) return defaults;

    final enabled = json['enabled'] != false;
    final rawTiers = json['tiers'];

    if (rawTiers is List && rawTiers.isNotEmpty) {
      final tiers = rawTiers
          .whereType<Map<String, dynamic>>()
          .map(ShippingTier.fromJson)
          .toList()
        ..sort((a, b) => a.minAmount.compareTo(b.minAmount));
      return ShippingSettings(enabled: enabled, tiers: tiers);
    }

    // Legacy threshold fields → ranges
    final baseCharge = _legacyDouble(json['baseCharge'], 250);
    final midTierThreshold = _legacyDouble(json['midTierThreshold'], 3000);
    final midTierCharge = _legacyDouble(json['midTierCharge'], 150);
    final freeShippingThreshold =
        _legacyDouble(json['freeShippingThreshold'], 6000);

    return ShippingSettings(
      enabled: enabled,
      tiers: [
        ShippingTier(
          minAmount: 0,
          maxAmount: (midTierThreshold - 1).clamp(0, double.infinity),
          charge: baseCharge,
        ),
        ShippingTier(
          minAmount: midTierThreshold,
          maxAmount: (freeShippingThreshold - 1).clamp(midTierThreshold, double.infinity),
          charge: midTierCharge,
        ),
        ShippingTier(
          minAmount: freeShippingThreshold,
          maxAmount: null,
          charge: 0,
        ),
      ],
    );
  }

  static double _legacyDouble(dynamic value, double fallback) {
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? fallback;
    return fallback;
  }
}

/// Calculate shipping from cart total (excl. GST) using vendor settings.
double calculateShippingCharges(
  double orderAmount, [
  ShippingSettings? settings,
]) {
  final cfg = settings ?? ShippingSettings.defaults;
  if (!cfg.enabled) return 0;

  final amount = orderAmount;
  for (final tier in cfg.tiers) {
    final minOk = amount >= tier.minAmount;
    final maxOk = tier.maxAmount == null || amount <= tier.maxAmount!;
    if (minOk && maxOk) return tier.charge;
  }

  if (cfg.tiers.isNotEmpty) return cfg.tiers.last.charge;
  return 0;
}

File: backend/utils/shipping.js
/**
 * Default shipping tiers (order amount range → shipping price).
 */
export const DEFAULT_SHIPPING_TIERS = [
  { minAmount: 0, maxAmount: 2999, charge: 250 },
  { minAmount: 3000, maxAmount: 5999, charge: 150 },
  { minAmount: 6000, maxAmount: null, charge: 0 },
];

export const DEFAULT_SHIPPING_SETTINGS = {
  enabled: true,
  tiers: DEFAULT_SHIPPING_TIERS,
};

const toNonNeg = (value, fallback) => {
  const n = Number(value);
  return Number.isFinite(n) && n >= 0 ? n : fallback;
};

/** Convert legacy threshold fields into range tiers. */
export const legacyToTiers = (raw = {}) => {
  const baseCharge = toNonNeg(raw.baseCharge, 250);
  const midTierThreshold = toNonNeg(raw.midTierThreshold, 3000);
  const midTierCharge = toNonNeg(raw.midTierCharge, 150);
  const freeShippingThreshold = toNonNeg(raw.freeShippingThreshold, 6000);

  return [
    {
      minAmount: 0,
      maxAmount: Math.max(0, midTierThreshold - 1),
      charge: baseCharge,
    },
    {
      minAmount: midTierThreshold,
      maxAmount: Math.max(midTierThreshold, freeShippingThreshold - 1),
      charge: midTierCharge,
    },
    {
      minAmount: freeShippingThreshold,
      maxAmount: null,
      charge: 0,
    },
  ];
};

const normalizeTier = (tiers) => {
  if (!Array.isArray(tiers) || tiers.length === 0) {
    return DEFAULT_SHIPPING_TIERS.map((t) => ({ ...t }));
  }

  return tiers
    .map((tier) => {
      const minAmount = toNonNeg(tier.minAmount, 0);
      const maxRaw = tier.maxAmount;
      const maxAmount =
        maxRaw === null || maxRaw === undefined || maxRaw === ''
          ? null
          : toNonNeg(maxRaw, null);
      const charge = toNonNeg(tier.charge, 0);
      return { minAmount, maxAmount, charge };
    })
    .sort((a, b) => a.minAmount - b.minAmount);
};

/**
 * Normalize vendor shippingSettings (fill defaults).
 * Supports new `tiers` and legacy threshold fields.
 */
export const normalizeShippingSettings = (raw = {}) => {
  const enabled = raw.enabled !== false;
  let tiers;

  if (Array.isArray(raw.tiers) && raw.tiers.length > 0) {
    tiers = normalizeTier(raw.tiers);
  } else if (
    raw.baseCharge != null ||
    raw.midTierThreshold != null ||
    raw.freeShippingThreshold != null
  ) {
    tiers = normalizeTier(legacyToTiers(raw));
  } else {
    tiers = DEFAULT_SHIPPING_TIERS.map((t) => ({ ...t }));
  }

  return { enabled, tiers };
};

/**
 * Calculate shipping from order amount + vendor settings.
 * @param {number} orderAmount - typically cart total (excl. GST)
 * @param {object} [settings] - vendor.shippingSettings
 * @returns {number}
 */
export const calculateShippingCharges = (orderAmount, settings) => {
  const cfg = normalizeShippingSettings(settings || {});
  if (!cfg.enabled) return 0;

  const amount = Number(orderAmount) || 0;
  const match = cfg.tiers.find((tier) => {
    const minOk = amount >= tier.minAmount;
    const maxOk = tier.maxAmount == null || amount <= tier.maxAmount;
    return minOk && maxOk;
  });

  if (match) return match.charge;

  // Fallback: use last tier if amount is above all ranges
  const last = cfg.tiers[cfg.tiers.length - 1];
  return last ? last.charge : 0;
};
